nav2_costmap_2d/include/nav2_costmap_2d/collision_footprint.hpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2020 Shivang Patel
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_COSTMAP_2D__COLLISION_FOOTPRINT_HPP_
+#define NAV2_COSTMAP_2D__COLLISION_FOOTPRINT_HPP_
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <algorithm>
+
+#include "rclcpp/rclcpp.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "geometry_msgs/msg/pose2_d.hpp"
+#include "nav2_costmap_2d/costmap_2d.hpp"
+#include "nav2_costmap_2d/collision_base.hpp"
+#include "nav2_costmap_2d/costmap_subscriber.hpp"
+#include "nav2_costmap_2d/footprint_subscriber.hpp"
+#include "nav2_util/robot_utils.hpp"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "tf2/utils.h"
+#pragma GCC diagnostic pop
+
+namespace nav2_costmap_2d
+{
+class CollisionFootprint : public CollisionBase
```

Comment:
Since this uses the collision tools, shouldn't all those be in the core collision checker? The external API of gtting the footprint cost seems to me to be a requirement for that.

Reply:
I'd suggest putting all this in the collision base class.

Reply:
I main reason it that original `CollisionChecker` classes uses exceptions. I don't want that in footprint checker

Reply:
Changing the inheritance and making some methods virtual overrides breaks ABI anyhow, if you're going to break it; break it

Reply:
Don't get me wrong here, but this design doesn't break `CollisionChecker`'s API (well at least I believe, and also I haven't modified its tests and they are still working and passing without any sweat).

If I have permission to break then let me see where its going to affect. Because the main thing I am worried about is - exceptions. I am worried that somewhere some code might rely on these exception will fail immediately. I think I should have investigated beforehand but that's a mistake on my part.

Anyways, let me see where `CollisionChecker` is used and then I will make the changes accordingly.

Reply:
I think this is very strange. No where in our stack this is being used. Even DWB has similar implementation of this in form of `obstacle_footprint` critic. I think we should remove this and make it purely non-subscription based.

Reply:
The subscriptions are used by the recoveries for their collision detection